contracts/NFTStorefrontV2.cdc

```diff
@@ -744,6 +765,30 @@ pub contract NFTStorefrontV2 {
             destroy listing
         }
 
+        /// cleanupGhostListings
+        /// Allow anyone to cleanup ghost listings
+        /// Listings will become ghost listings if stored provider capability doesn't hold
+        /// the NFT anymore.
+        ///
+        /// @param listingResourceID ID of the listing resource which would get removed if it become ghost liting.
+        pub fun cleanupGhostListings(listingResourceID: UInt64) {
+            pre {
+                self.listings[listingResourceID] != nil: "Could not find listing with given id"
+            }
+            let listingRef = self.borrowListing(listingResourceID: listingResourceID)!
+            let details = listingRef.getDetails()
+            assert(!details.purchased, message: "Given listing is already purchased")
```

Comment:
Why are we panicking here instead of a no-op? Is it easier for a marketplace frontend to determine state on a reversion than a no-op?

Reply:
If I would make it no-op then marketplaces have to explicitly check whether the listing got deleted. And there would be a high chance that marketplaces may miss that piece of the code from their transaction, and I always think that it is better to provide the revert in the transaction instead of no-op as the end user would never know the reason for the no-op if I don't do a revert.

Reply:
Why can't I cleanup purchased listings? Now I would have to make two different function calls to clean a wallet's storage

Reply:
Why do you need to make two calls? You can use `cleanupPurchasedListing` function during sell_item transaction while use `cleanupGhostListing` when cleaning ghost listing. Both will not occur at the same time. It is obvious that marketplaces would use `cleanupPurchasedListing` in sell item transactions.

Reply:
I also think it should be a no-op